### ISSUE TITLE
Data saving

### ISSUE THREAD

**Issue by socialdecisionlab** [nataliavelez/playable_lines#9]
Figure out data saving:
Save parameters to player, run, stage in empirica
Currently saving x,y location of each player in round.
Need:
What parameters we want to save
Which level (player, players, round, stage and so on) to save them

**Comment by WhiteJP** [comment]
Hi team, 

This looks like it might be a more complicated issue than I first imagined. I had thought (maybe wishfully hoped), that empirica took care of keeping track of every time the state changes and saves it along the way when you export the data. But alas, it appears that it doesn't -- it only has the final state of an attribute. There's plenty of possible solutions, but will be more intensive than thought. For a bit more context here's what I wrote on the empirica slack to see if i could get some advice:

---
Hi team. I'm running an experiment that is a multi-round 8-player multiplayer 2d grid game where people walk around and do stuff etc. I need to save the data of every position move of every player (and more). Is there an elegant way to do this?

1. having a 'data-stream' round attribute that, every time the shared state changes, I just append the updated data. This is probably the easiest solution, but I'm worried the data will be very long, and I'll run into this error that 
[@Ham Huang](https://empirica-ly.slack.com/team/U06RXT8MEMN)
 and 
[@Mohammed Alsobay](https://empirica-ly.slack.com/team/U023QKU9U1W)
 are talking about above.

2. round.set() an attribute each time state changes, "state1", "state2" etc and saving it to the round that way. This might end up with some pretty wide data but might also work.

Wondering if anybody has any thoughts about these (might one be the better choice than the other), or if there's another more elegant way to do this in empirica?

(3. maybe managing a database elsewhere and just sending it off each time state changes?)

**Comment by nataliavelez** [comment]
Ham’s working on a solution to this very issue now - I think what’s preventing him from downloading the full mouse tracking data is actually just a config error, so Mohammed fixed it pretty quickly. I’ll let you know once that’s working!

—

Sent from my phone
________________________________

**Comment by WhiteJP** [comment]
Yeah, I've been speaking with Ham. I think one needs to build empirica from scratch from github (which requires installing go, and doesn't have instructions haha) to change that one little line of config though :/ will do that if it's the best approach, but I'm hoping there might be another way. We'll see :)

**Comment by WhiteJP** [comment]
So after speaking to Ham and Mohammed, plan is:

- [x] try easiest option just append new player states each time it is updated, see if that is too long for current empirica
- [ ] If it is, might be a good idea to save the data in a more advantageous way. First, maybe try removing things from playerState that don't change within the round, like name, color, and use player number as the key, not the super long empirica generated id).  If that doesn't work can try storing `x`, `y`, `carrying`, `score`, all as separate attributes on player.round and appending (this will likely work under the size limit). Will likely do this anyway as part of improving syncrhonization (https://github.com/nataliavelez/playable_lines/issues/18)

**Comment by nataliavelez** [comment]
Sounds like a solid plan - thanks for the update!

—

Sent from my phone
________________________________

**Comment by WhiteJP** [comment]
FYI, empirica have released a new version (`v1.11.3`) which should fix this buffer issue. To upgrade run `empirica upgrade` in your empirica project folder.

**Comment by WhiteJP** [comment]
Also, this issue only relates to the `empirica export` command which exports `tajriba.json` into a csv. Regardless of whether we can get this to work or not, we still have access to all the data and the updates in `tajriba.json`, and could easily write our own function to extract relevant data from this.  So regardless, isn't the biggest issue.

**Comment by WhiteJP** [comment]
New version of empirica largely solved this problem. Currently we save all changes to the shared state and their timestamp, and append it to a large array. If this becomes a problem, we don't need to do this and can just work directly from the server json file `tajriba.json`, which saves everything along the way.
